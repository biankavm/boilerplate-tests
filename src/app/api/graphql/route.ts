import { NextRequest, NextResponse } from 'next/server'

let tasks = [
  {
    id: '1',
    title: 'Tarefa exemplo 1',
    completed: false,
    createdAt: new Date().toISOString()
  },
  {
    id: '2',
    title: 'Tarefa exemplo 2',
    completed: true,
    createdAt: new Date().toISOString()
  }
]

let users = [
  {
    id: '1',
    name: 'Jo√£o Silva',
    email: 'joao@example.com',
    createdAt: new Date().toISOString()
  },
  {
    id: '2',
    name: 'Maria Santos',
    email: 'maria@example.com',
    createdAt: new Date().toISOString()
  }
]

function executeGraphQL(query: string, variables: Record<string, unknown>) {
  // GET_TASKS
  if (query.includes('query GetTasks')) {
    return {
      data: {
        tasks
      }
    }
  }

  // GET_USER
  if (query.includes('query GetUser')) {
    const user = users.find((u) => u.id === variables.id)
    return {
      data: {
        user: user || null
      }
    }
  }

  // CREATE_TASK
  if (query.includes('mutation CreateTask')) {
    const title = typeof variables.title === 'string' ? variables.title : ''
    const newTask = {
      id: String(tasks.length + 1),
      title,
      completed: false,
      createdAt: new Date().toISOString()
    }
    tasks.push(newTask)
    return {
      data: {
        createTask: newTask
      }
    }
  }

  // TOGGLE_TASK
  if (query.includes('mutation ToggleTask')) {
    const task = tasks.find((t) => t.id === variables.id)
    if (task) {
      task.completed = !task.completed
      return {
        data: {
          toggleTask: {
            id: task.id,
            completed: task.completed
          }
        }
      }
    }
    return {
      errors: [{ message: 'Task not found' }]
    }
  }

  // DELETE_TASK
  if (query.includes('mutation DeleteTask')) {
    const index = tasks.findIndex((t) => t.id === variables.id)
    if (index !== -1) {
      tasks.splice(index, 1)
      return {
        data: {
          deleteTask: {
            id: variables.id
          }
        }
      }
    }
    return {
      errors: [{ message: 'Task not found' }]
    }
  }

  return {
    errors: [{ message: 'Unknown query or mutation' }]
  }
}

export async function POST(request: NextRequest) {
  try {
    const body = await request.json()
    const { query, variables } = body

    const result = executeGraphQL(query, variables || {})

    return NextResponse.json(result)
  } catch {
    return NextResponse.json(
      {
        errors: [{ message: 'Invalid request' }]
      },
      { status: 400 }
    )
  }
}
