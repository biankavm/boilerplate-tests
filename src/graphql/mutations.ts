import { gql } from '@apollo/client'

export const CREATE_TASK = gql`
  mutation CreateTask($title: String!) {
    createTask(title: $title) {
      id
      title
      completed
      createdAt
    }
  }
`

export const TOGGLE_TASK = gql`
  mutation ToggleTask($id: ID!) {
    toggleTask(id: $id) {
      id
      completed
    }
  }
`

export const DELETE_TASK = gql`
  mutation DeleteTask($id: ID!) {
    deleteTask(id: $id) {
      id
    }
  }
`
