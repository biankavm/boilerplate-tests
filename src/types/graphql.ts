export interface Task {
  id: string
  title: string
  completed: boolean
  createdAt: string
}

export interface User {
  id: string
  name: string
  email: string
  createdAt: string
}

export interface GetTasksData {
  tasks: Task[]
}

export interface GetUserData {
  user: User | null
}

export interface GetUserVariables {
  id: string
}
