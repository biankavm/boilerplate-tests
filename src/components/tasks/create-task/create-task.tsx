'use client'

import { useState } from 'react'
import { useMutation } from '@apollo/client/react'
import { CREATE_TASK, GET_TASKS } from '@/graphql'
import styles from './create-task.module.scss'

export function CreateTask() {
  const [title, setTitle] = useState('')
  const [createTask, { loading, error }] = useMutation(CREATE_TASK, {
    refetchQueries: [{ query: GET_TASKS }],
    onCompleted: () => {
      setTitle('')
    }
  })

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault()
    if (title.trim()) createTask({ variables: { title: title.trim() } })
  }

  return (
    <form onSubmit={handleSubmit} className={styles.form}>
      <input
        type="text"
        value={title}
        onChange={(e) => setTitle(e.target.value)}
        placeholder="Digite uma nova tarefa..."
        disabled={loading}
        className={styles.input}
        aria-label="TÃ­tulo da nova tarefa"
      />

      <button
        type="submit"
        disabled={loading || !title.trim()}
        className={styles.button}
        aria-label="Criar nova tarefa"
      >
        {loading ? 'Criando...' : 'Criar'}
      </button>
      {error && (
        <p className={styles.error} role="alert">
          Erro ao criar tarefa: {error.message}
        </p>
      )}
    </form>
  )
}
