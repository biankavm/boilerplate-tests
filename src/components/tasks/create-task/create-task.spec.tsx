import { CreateTask } from './create-task'
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { CREATE_TASK } from '@/graphql'
import { MockedProvider } from '@apollo/client/testing/react'

const createTaskMock = (title: string) => ({
  request: {
    query: CREATE_TASK,
    variables: { title }
  },
  result: {
    data: {
      createTask: {
        id: '1',
        title,
        completed: false,
        createdAt: '2024-01-01T00:00:00.000Z'
      }
    }
  }
})

describe('CreateTask', () => {
  const renderComp = ({ title = 'my task' }: { title?: string } = {}) => {
    const mocks = title ? [createTaskMock(title)] : []

    return render(
      <MockedProvider mocks={mocks}>
        <CreateTask />
      </MockedProvider>
    )
  }

  it('should render correctly', () => {
    const { container } = renderComp()
    expect(container).toMatchSnapshot()
  })

  it('should have a correctly placeholder in input', () => {
    renderComp()

    expect(
      screen.getByPlaceholderText(/digite uma nova tarefa/i)
    ).toBeInTheDocument()
  })

  it('should create task when filling input and clicking button', async () => {
    const taskTitle = 'Nova tarefa'

    renderComp({ title: taskTitle })

    const user = userEvent.setup() // criando instância de usuário (com contexto próprio)

    const input = screen.getByPlaceholderText(/digite uma nova tarefa/i)
    const button = screen.getByRole('button', { name: /criar/i })

    expect(button).toBeDisabled() // botão começa desabilitado

    // simula o usuário preenchendo o input
    await userEvent.type(input, taskTitle)

    expect(button).not.toBeDisabled() // após o usuário digitar, habilitamos

    expect(input).toHaveValue(taskTitle) // após o usuário digitar, o input tem o valor da task

    await user.click(button)
    expect(button).toHaveTextContent('Criando...') // botão muda para "Criando..."

    // após criar a task, o input é limpo
    // waitFor: é um método que espera uma condição ser true
    await waitFor(() => {
      expect(input).toHaveValue('')
      expect(button).toHaveTextContent('Criar')
    })
  })
})
