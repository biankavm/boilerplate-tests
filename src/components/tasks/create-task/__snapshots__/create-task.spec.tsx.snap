// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CreateTask should render correctly 1`] = `
<div>
  <form
    class="form"
  >
    <input
      aria-label="TÃ­tulo da nova tarefa"
      class="input"
      placeholder="Digite uma nova tarefa..."
      type="text"
      value=""
    />
    <button
      aria-label="Criar nova tarefa"
      class="button"
      disabled=""
      type="submit"
    >
      Criar
    </button>
  </form>
</div>
`;
